### PR TITLE
z_en_kakasi3.c OK

### PR DESCRIPTION
**Pull request by i82orbom**
The last round on the kakasi series.

To whoever reviews this PR, there's other z_en_kakasiX.c PRs open, they're very similar in content as there was a lot of copy & pasta

### CODE REVIEW

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:365]: this is where I was writing about an early return making more sense

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:378]:
```suggestion
            if (!(gSaveContext.eventChkInf[9] & 0x1000)) {
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:389]:
```suggestion
        globalCtx->msgCtx.unk_E3EE = 4;
```
there is another one earlier in the file, unless they're necessary?

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:410]: You have `s16 unk_19C[0x3];` (notice the length, 3, vs this for loop assuming 4), could `unk_1A2` actually be part of that array instead of being on its own?

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:406]:
```suggestion
    this->unk_198++;
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:427]:
```suggestion
    SkelAnime_DrawSV(globalCtx, this->skelAnime.skeleton, this->skelAnime.limbDrawTbl, this->skelAnime.dListCount, NULL, NULL,
                     &this->actor);
```
pass `NULL` for the function pointers callbacks

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h:27]:
```suggestion
    /* 0x01AA */ s16 unk_1AA;
    /* 0x01AC */ char unk_1AC[2];
    /* 0x01AE */ s16 unk_1AE;
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h:33]:
```suggestion
    /* 0x020A */ char unk_20A[0x2];
```
just for consistency with the other unk_xxx names in the struct

**Review by Dragorn421 — CHANGES_REQUESTED** [review]

**Comment by Dragorn421** [comment]
Glossing over #428 very quickly it does look very similar to this so I'll let you apply the review comments to it too (and any other contributions) :)

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:62]: This string is very likely related to the name of the scarecrow, aka "Bonooru"

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:225]: In regards to the absAngle, unfortunately it doesn't match if I do it this way. I've tried many different options and they don't match...

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:222]: Yep, it doesn't match 😞

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:342]: Yep, the flow of this block was a bit weird, I've changed it to else if and merged the conditions, let me know what you think

**Comment by i82orbom** [comment]
@Roman971 @Dragorn421 Thank you for the PR review, I've applied your suggestions to this PR and to https://github.com/zeldaret/oot/pull/428 and https://github.com/zeldaret/oot/pull/427 as the three actors are closely related.

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h:33]: probably padding
```suggestion
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.h:28]: also probably padding
```suggestion
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:8]:
```suggestion
#include "vt.h"
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:79]: Does something like this work by any chance? Might have to reorder things to keep it matching
```suggestion
    this->unk_1A4 = 0;
    this->skelAnime.animPlaybackSpeed = 0.0f;
    this->unk_1AA = this->unk_1AE = 0;
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:141]:
```suggestion
        Math_SmoothScaleMaxF(&this->skelAnime.animPlaybackSpeed, this->unk_1B8, 0.1f, 0.2f);
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:182]: Based on existing usage I don't think we can really tell whether this is a boolean.
```suggestion
        if (gSaveContext.unk_12C5 != 0) {
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:189]:
```suggestion
        if (gSaveContext.unk_12C5 != 0) {
```

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:190]: This one is treated as a boolean here but not in the rest of the file. I'm not sure what it should actually be but I think it should be consistent at least.

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:237]: No need for the else if there is already an early return imo

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:252]: Same here about the else + early return

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:316]:
```suggestion
        globalCtx->msgCtx.unk_E3EE = 4;
```

**Review by Roman971 — COMMENTED** [review]
A few mostly minor things. Some are probably relevant in the other En_Kakasi PRs

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:237]: These two early returns are necessary, otherwise the flow is incorrect, notice the call to func_8002F2CC down below.

Comment (Roman971) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:236]: Right, but I'm not saying the early returns should go away, only the `else`s

Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c]: Oh, I misunderstood, you're right

**Review by Roman971 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:115]: `unk_1AE` and `unk_1AA` are used as rotation target values so hexadecimal should be used, even if the decimal values are nicer.

Same for parameters in calls such as `Math_SmoothScaleMaxMinS(&this->actor.shape.rot.x`, the numeric arguments should be written in hex when it's about rotations

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:356]:
```suggestion
        this->unk_195 = true;
```
Since you're using it as a bool everywhere

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:180]: same as `unk_195`, and a few lines below too

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:293]: early return instead of an else here imo, for consistency with other early returns after action setups

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:275]: You didn't merge those ifs but the next one `if (globalCtx->msgCtx.unk_E3EE == 3 && globalCtx->msgCtx.msgMode == 0) {` is merged, probably do one or the other but not both for consistency

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:348]: same comments as before about early returns and merged ifs in this function `func_80A918E4`

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:404]:
```suggestion
    for (i = 0; i < ARRAY_COUNT(this->unk_19C); i++) {
```

**Review by Dragorn421 — CHANGES_REQUESTED** [review]

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:275]: you're right, I'll keep it unmerged, imo is a bit more readable this way...

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:299]: no need to return; at the end of the function, and also `func_80A90EBC` isn't an action setup

sorry if I'm being annoying with this 😅

**Review by Dragorn421 — CHANGES_REQUESTED** [review]

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (fig02) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:299]: cleaning up stuff like this at this stage is good, needs to be done :D
after this change this is good to for me

**Review by fig02 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c:299]: don't worry it's fine, it's hard to keep apply the suggestions across all 3 PR's 😅 

What does imply that the function isn't a action setup?

Comment (Dragorn421) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c]: What I call "action setup" is anything that does `this->actionFunc = ...;`, often along with setting a bunch of struct members
`func_80A90EBC` doesn't do that anywhere, its purpose seems to be reusing common logic from different actions but that's up to further documentation

Unrelated but avoid force-pushing please, it ruins tracking of comments and commits by github

**Review by Dragorn421 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c
Comment (i82orbom) [reply src/overlays/actors/ovl_En_Kakasi3/z_en_kakasi3.c]: Ah, ok thanks for the clarification.

Yeah I've just noticed about the force push

**Pull request merged**
